File: src/main/java/org/intellij/sdk/language/jass/JassStructureViewElement.java
package org.intellij.sdk.language.jass;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.navigation.ItemPresentation;
import com.intellij.psi.NavigatablePsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import org.intellij.sdk.language.jass.psi.JassFile;
import org.intellij.sdk.language.jass.psi.JassProperty;
import org.intellij.sdk.language.jass.psi.impl.JassPropertyImpl;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class JassStructureViewElement implements StructureViewTreeElement, SortableTreeElement {

    private final NavigatablePsiElement myElement;

    public JassStructureViewElement(NavigatablePsiElement element) {
        this.myElement = element;
    }

    @Override
    public Object getValue() {
        return myElement;
    }

    @Override
    public void navigate(boolean requestFocus) {
        myElement.navigate(requestFocus);
    }

    @Override
    public boolean canNavigate() {
        return myElement.canNavigate();
    }

    @Override
    public boolean canNavigateToSource() {
        return myElement.canNavigateToSource();
    }

    @NotNull
    @Override
    public String getAlphaSortKey() {
        String name = myElement.getName();
        return name != null ? name : "";
    }

    @NotNull
    @Override
    public ItemPresentation getPresentation() {
        ItemPresentation presentation = myElement.getPresentation();
        return presentation != null ? presentation : new PresentationData();
    }

    @Override
    public TreeElement @NotNull [] getChildren() {
        if (myElement instanceof JassFile) {
            List<JassProperty> properties = PsiTreeUtil.getChildrenOfTypeAsList(myElement, JassProperty.class);
            List<TreeElement> treeElements = new ArrayList<>(properties.size());
            for (JassProperty property : properties) {
                treeElements.add(new JassStructureViewElement((JassPropertyImpl) property));
            }
            return treeElements.toArray(new TreeElement[0]);
        }
        return EMPTY_ARRAY;
    }

}

File: src/main/kotlin/raft/war/language/jass/structureView/JassStructureViewElement.kt
package raft.war.language.jass.structureView

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement
import com.intellij.ide.util.treeView.smartTree.TreeElement
import com.intellij.navigation.ItemPresentation
import com.intellij.psi.NavigatablePsiElement
import com.intellij.psi.util.PsiTreeUtil
import raft.war.language.jass.JassIcons
import raft.war.language.jass.psi.JassFun
import raft.war.language.jass.psi.JassNativ
import raft.war.language.jass.psi.file.JassFile
import raft.war.language.jass.psi.impl.JassFunImpl
import raft.war.language.jass.psi.impl.JassNativImpl
import javax.swing.Icon


class JassStructureViewElement(private val element: NavigatablePsiElement) : StructureViewTreeElement,
    SortableTreeElement {
    override fun getValue(): Any = element

    override fun navigate(requestFocus: Boolean) = element.navigate(requestFocus)

    override fun canNavigate(): Boolean = element.canNavigate()

    override fun canNavigateToSource(): Boolean = element.canNavigateToSource()

    override fun getAlphaSortKey(): String {
        val name = element.name
        return name ?: ""
    }

    override fun getPresentation(): ItemPresentation {
        val presentation = element.presentation
        return presentation ?: object : PresentationData() {
            override fun getPresentableText(): String {
                if (element is JassFun) {
                    return element.funHead?.funName?.text ?: "Undefined"
                }
                if (element is JassNativ) {
                    return element.funHead?.funName?.text ?: "Undefined"
                }
                return "Undefined"
            }

            override fun getIcon(open: Boolean): Icon? {
                if (element is JassFun) return JassIcons.function
                if (element is JassNativ) return JassIcons.native
                return null
            }

            override fun getLocationString(): String = element.containingFile.name
        }
    }

    override fun getChildren(): Array<TreeElement> {
        val treeElements: MutableList<TreeElement> = mutableListOf()

        if (element is JassFile) {
            val items = PsiTreeUtil.getChildrenOfAnyType(element, JassFun::class.java, JassNativ::class.java)
            for (item in items) {
                if (item is JassFun) treeElements.add(JassStructureViewElement(item as JassFunImpl))
                if (item is JassNativ) treeElements.add(JassStructureViewElement(item as JassNativImpl))
            }
        }
        return treeElements.toTypedArray<TreeElement>()
    }
}
